File: fsharp-csharp-interop-demos/unit-testing-demo/Domain.Tests.CSharp/FlowerServiceTests.cs
using FluentAssertions;
using NSubstitute;
using Xunit;

namespace Domain.Tests.CSharp
{
    public class FlowerServiceTests
    {
        [Fact]
        public void Send_happy_case_calls_deliveryService()
        {
            // Arrange
            var deliveryServiceMock = Substitute.For<IDeliveryService>();
            var sut = new FlowerService(deliveryServiceMock);

            // Act
            sut.Send(new Order{ From = "Bob", To = "Alice", FlowerName = "Rose" });

            // Assert
            deliveryServiceMock.Received().SendFlowers(Arg.Any<Order>());
        }

        [Fact]
        public void Send_happy_case_returns_success()
        {
            // Arrange
            var deliveryServiceMock = Substitute.For<IDeliveryService>();
            deliveryServiceMock.SendFlowers(Arg.Any<Order>())
                .Returns(new SendResponse {WasSuccessfull = true});

            var sut = new FlowerService(deliveryServiceMock);

            // Act
            var result = sut.Send(new Order{ From = "Bob", To = "Alice", FlowerName = "Rose" });

            // Assert
            result.WasSuccessfull.Should().BeTrue();
        }
    }
}

File: fsharp-csharp-interop-demos/unit-testing-demo/Domain.Tests.FSharp/FlowerServiceTests.fs
module FlowerServiceTests

open System
open Foq
open FsUnit.Xunit
open Xunit
open Domain
open NHamcrest

// http://brandewinder.com/2013/01/27/Testing-and-mocking-your-C-sharp-code-with-F-sharp/
[<Fact>]
let ``Send happy case calls deliveryService`` () =

    // TODO: Show manual mock before using FOQ library

    // Arrange
    let deliveryServiceMock =
        Mock<IDeliveryService>()
            .Setup(fun x -> <@ x.SendFlowers(any()) @>)
            .Returns(SendResponse(WasSuccessfull = true))
            .Create();

    let sut = FlowerService(deliveryServiceMock);

    // Act
    let result = sut.Send(Order( From = "Bob", To = "Alice", FlowerName = "Rose" ));

    // Assert
    Assert.True(result.WasSuccessfull)
